runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkSubmissionTest.java

```diff
@@ -0,0 +1,224 @@
+      // Run end-to-end test
+      CliFrontend.main(argsBuilder.build().toArray(new String[0]));
+    } catch (SystemExitException e) {
+      // The Clifrontend exited and we can move on to check if the job has finished
+    } finally {
+      restoreDefaultSystemExitBehavior();
+    }
+
+    waitUntilJobIsCompleted();
+  }
+
+  /** The Flink program which is executed by the CliFrontend. */
+  public static void main(String[] args) {
+    FlinkPipelineOptions options = PipelineOptionsFactory.as(FlinkPipelineOptions.class);
+    options.setRunner(FlinkRunner.class);
+    options.setParallelism(1);
+    Pipeline p = Pipeline.create(options);
+    p.apply(GenerateSequence.from(0).to(1));
+    p.run();
+  }
+
+  private static void prepareEnvironment() throws Exception {
+    // Write a Flink config
+    File file = TEMP_FOLDER.newFile("flink-conf.yaml");
+    String config =
+        String.format(
+            "%s: %s\n%s: %s\n%s: %s",
+            JobManagerOptions.ADDRESS.key(),
+            "localhost",
+            JobManagerOptions.PORT.key(),
+            flinkCluster.getClusterPort(),
+            RestOptions.PORT.key(),
+            flinkCluster.getRestPort());
+    Files.write(file.toPath(), config.getBytes(Charsets.UTF_8));
+
+    // Create a new environment with the location of the Flink config for Clifrontend
+    ImmutableMap<String, String> newEnv =
+        ImmutableMap.<String, String>builder()
+            .putAll(ENV.entrySet())
+            .put(ConfigConstants.ENV_FLINK_CONF_DIR, file.getParent())
+            .build();
+
+    modifyEnv(newEnv);
+  }
+
+  private static void restoreEnvironment() throws Exception {
+    modifyEnv(ENV);
+  }
+
+  private static void modifyEnv(Map<String, String> env) throws Exception {
+    Class processEnv = Class.forName("java.lang.ProcessEnvironment");
+    Field envField = processEnv.getDeclaredField("theUnmodifiableEnvironment");
+
+    Field modifiersField = Field.class.getDeclaredField("modifiers");
+    modifiersField.setAccessible(true);
+    modifiersField.setInt(envField, envField.getModifiers() & ~Modifier.FINAL);
+
+    envField.setAccessible(true);
+    envField.set(null, env);
+    envField.setAccessible(false);
+
+    modifiersField.setInt(envField, envField.getModifiers() & Modifier.FINAL);
+    modifiersField.setAccessible(false);
+  }
+
+  private void waitUntilJobIsCompleted() throws Exception {
+    while (true) {
+      Collection<JobStatusMessage> allJobsStates = flinkCluster.listJobs().get();
+      assertThat(
+          "There should be a job per test run.", allJobsStates.size(), is(expectedNumberOfJobs));
+      if (allJobsStates.stream()
+          .allMatch(jobStatus -> jobStatus.getJobState() == JobStatus.FINISHED)) {
+        return;
+      }
+      Thread.sleep(50);
+    }
+  }
+
+  /** Prevents the CliFrontend from calling System.exit. */
+  private static void throwExceptionOnSystemExit() {
+    System.setSecurityManager(
+        new SecurityManager() {
```

Comment:
Instead of creating an all-new SecurityManager, should we instead wrap the original one? (Not sure if it actually matters.)

Reply:
I don't think it matters for tests, but doesn't hurt either. Updated.